AX: assert if AXIsolatedTree::applyPendingRootNodeLocked results in unreachable nodes

minorninth opened the pull request:
#### 1530f996d732b1c894c62eb91726a9689e62f506
<pre>
AX: assert if AXIsolatedTree::applyPendingRootNodeLocked results in unreachable nodes
<a href="https://bugs.webkit.org/show_bug.cgi?id=296167">https://bugs.webkit.org/show_bug.cgi?id=296167</a>
<a href="https://rdar.apple.com/156121932">rdar://156121932</a>

Reviewed by Tyler Wilcock and Andres Gonzalez.

We have tracked down the source of one missing content bug to
AXIsolatedTree::applyPendingRootNodeLocked setting the root node to an
iframe&apos;s scroll view by mistake. This assertion would catch if that
were to happen again.

* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
(WebCore::AXIsolatedTree::applyPendingRootNodeLocked):

Canonical link: <a href="https://commits.webkit.org/297632@main">https://commits.webkit.org/297632@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1b9400e58e7bff3bc8241b258bf81bff8a56d95e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/112314 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/32045 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/22523 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/118393 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/62712 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/114276 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/32698 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/40609 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/85320 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/36047 "Build was cancelled. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running run-layout-tests-in-stress-mode; layout-tests (failure); Uploaded test results; layout-tests (failure); Compiled WebKit (cancelled)") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/115261 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/26086 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/101049 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/65751 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/25389 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/19186 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/62238 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/95470 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/19264 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/121718 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/39388 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/29311 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/94130 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/39769 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/97291 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/93955 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24033 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/39193 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/16991 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/35437 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/39276 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/44764 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/38911 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/42248 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/40654 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

AndresGonzalezApple commented on Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
```diff
@@ -190,6 +190,21 @@ void AXIsolatedTree::applyPendingRootNodeLocked()
+                    self(&child.get(), reachableNodes, self);
+            };
+            HashSet<AXID> reachableNodes;
+            if (!!m_rootNode) {
```
Why !!m_rootNode instead of m_rootNode?

minorninth replied:
Fixed, thanks

twilco commented on Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
```diff
@@ -190,6 +190,21 @@ void AXIsolatedTree::applyPendingRootNodeLocked()
+                markReachableNodes(m_rootNode.get(), reachableNodes, markReachableNodes);
+                if (reachableNodes.size() != m_nodeMap.size())
+                    AXLOG(makeString("AX: After applying pending root node, %u reachable nodes but %u are in the node map"_s, reachableNodes.size(), m_nodeMap.size()));
+                ASSERT(reachableNodes.size() == m_nodeMap.size());
```
I've personally made a habit of not using AXLOG. There are too many of them all over the place such that they are really spammy, and requires `LOG_DISABLED` to be set to `0`, which basically requires a debug build. Whereas I've had success building release with `ASSERT_ENABLED`.

Maybe the answer is to clean up the spammy AXLOGs and to profile our debug builds so they aren't unusably slow on most websites, then truly live on debug builds. But I wonder if for now, we could use ASSERT_WITH_MESSAGE for this instance?

 `ASSERT_WITH_MESSAGE(reachableNodes.size() == m_nodeMap.size(), "expected %u reachable nodes, but only had %u after applying the root node", m_nodeMap.size(), reachableNodes.size());`
 
Also something I just realized — technically, `m_nodeMap` is strictly supposed to only be used on the main-thread, while this method runs off the main-thread. So things like calling `size()` on this hash map while the main-thread is the middle of a rehash (or something) is possible, which would cause a crash. But seems very unlikely, and this check would be rarely run, so might be fine in order to get this nice assert.

minorninth replied:
Good idea, switched to ASSERT_WITH_MESSAGE. I also switched to m_readerThreadNodeMap which is supposed to be used from the secondary thread.

minorninth committed:
AX: assert if AXIsolatedTree::applyPendingRootNodeLocked results in unreachable nodes
https://bugs.webkit.org/show_bug.cgi?id=296167
rdar://156121932

Reviewed by Tyler Wilcock and Andres Gonzalez.

We have tracked down the source of one missing content bug to
AXIsolatedTree::applyPendingRootNodeLocked setting the root node to an
iframe's scroll view by mistake. This assertion would catch if that
were to happen again.

* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
(WebCore::AXIsolatedTree::applyPendingRootNodeLocked):

Canonical link: https://commits.webkit.org/297632@main
